public/app/features/query/components/QueryEditorRow.tsx

```diff
@@ -565,7 +565,12 @@ export function filterPanelDataToQuery(data: PanelData, refId: string): PanelDat
 
   // Only say this is an error if the error links to the query
   let state = data.state;
+  // This is an error in the *previous* query, not the new one.
+  // (Both results are being merged in runRequest/preProcessPanelData() which results in a combination of data)
   const error = data.error && data.error.refId === refId ? data.error : undefined;
+  // An alternative solution is in the comment below, but I feel like it's more dealing with the symptoms and not the cause?
+  // Unless of course that the error here is persisted from the last query, is intended behaviour.
+  // if (state !== LoadingState.Loading && error) {
```

Comment:
hey @idastambuk is your proposal to add `if (state !== LoadingState.Loading && error) {` And the runRequest change below, or is this OR that?

Reply:
It's either or! I updated the description to be more clear :)

Reply:
> Hey @idastambuk I think the thing I'm missing here is why the loading state never resolves the second time around? I think it's a fair question whether or not we clear an error when we make the request the second time around or when we get a non-errored response, but I think regardless it feels like we should get the loading/not loading state passed down to us in someway, and I think I'm missing where that happens/doesn't happen, you know?

Definitely. The explanation is a bit long but: the reason we dont clear the loading state the second time around has to do with the RunQueryButtons component in Redshift, which keeps 2 internal loading states ("cancelling" and "running" query) and updates them according to the LoadingState it gets from the panel. 

In this case, however, the state is kept in LoadingState.Error (because of the above logic) and the component never triggers the useEffect that tells it to stop loading. 

The place where the error state is persisted is just before where QueryEditor is being imported, in  QueryGroupRow: 

```
if (error) {
    state = LoadingState.Error; //so even if there's a loading state, if the error is truthy, it will overwrite it

```

If it gets a different response, like Success, it will update. Meaning the only use case where this bug appears is if there are two errors in a row. AND if any of the plugin components depend on this loading state. Which they do when they implement long-running features. (The same thing happens in Athena.) 

Of course then it might be said: Why are we fixing it in Grafana repo and not the buttons component, but even if we want to fix it and only depend on the state from the panel, we need the correct state actually passed down (as you said)

Reply:
Ah ok so I think this is starting to make sense to me! Going to repeat it back and you can correct me if I'm wrong:
- we currently override the loading state in this file to "Error" whenever we encounter an error
- we always encounter an error if the last request had an error when we "preProcessPanelData" below we never remove old errors from the panel data, we simply keep whatever lastResult was. 

- We we decide to fix "preProcessPanelData" to clear out old errors we will get up to date loading state data. However we may cause other datasources to be a bit "jumpy" if a user refreshes a broken query and gets an error, clears it out, and then gets the same error popping up again
- An alternative to this jumpy behavior is to not override the state here to an error state when we are still loading. The pros are that this would not cause jumpy-ness, the cons are that we wouldn't know we were in a loading state until all the queries had finished getting called.

Does that sound right? Definitely seems worth talking about with other developers outside of team to get their thoughts! I think I can see an argument either way. I think I would lean towards fixing it here in QueryEditorRow so that we pass along the loading state and don't manually override to error until loading is done. It seems less disruptive maybe? But I would be curious to see what others think!

I am still unclear why this has only become an issue after releasing long running queries, it feels like this would have existed for lots of datasources? Even when we didn't have a run query button in the query editor we still have it outside? Maybe we do weird logic to handle it there?